File: kotlin/crypto-square/src/main/kotlin/CryptoSquare.kt
import java.util.*
import kotlin.math.ceil
import kotlin.math.sqrt

object CryptoSquare {
    fun ciphertext(plaintext: String) =
        if (plaintext.isEmpty()) "" else plaintext.normalize().rows().columns().joinToString(" ")

    private fun String.normalize() = filter(Char::isLetterOrDigit).lowercase(Locale.getDefault())
    private fun String.size() = ceil(sqrt(length.toDouble())).toInt()
    private fun String.rows() = size().let { size -> chunked(size).map { it.padEnd(size) } }
    private fun List<String>.columns() = this[0].indices.map { i -> this.map { it[i] }.joinToString("") }
}

File: fsharp/crypto-square/CryptoSquare.fs
module CryptoSquare

open System

let private normalize (plaintext: string) =
    plaintext.ToLower() |> String.filter Char.IsLetterOrDigit
    
let private size (plaintext: string) = int (Math.Ceiling(Math.Sqrt(float plaintext.Length)))

let private rows (plaintext: string) =
    let n = size plaintext

    plaintext.ToCharArray()
    |> Array.chunkBySize (max 1 n)
    |> Array.map (fun row -> Array.append row (Array.replicate (n - row.Length) ' '))

let ciphertext (plaintext: string) = 
    plaintext
    |> normalize
    |> rows
    |> Array.transpose
    |> Array.map String 
    |> String.concat " "
